File: April2022/Java/Shift2DGrid.java

// Author: Shobhit Behl (LC: shobhitbruh)
class Solution {
    public List<List<Integer>> shiftGrid(int[][] grid, int k) {
        int n=grid.length;
        int m=grid[0].length;
        
        LinkedList<Integer> li=new LinkedList<>();
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                li.addLast(grid[i][j]);
            }
        }
        while(k-->0){
        int f=li.removeLast();
        li.addFirst(f);
        }
        List<List<Integer>> al=new ArrayList<>();
        for(int i=0;i<n;i++){
            List<Integer> g=new ArrayList<>();
            for(int j=0;j<m;j++){
               g.add(li.removeFirst()); 
            }
            al.add(g);
        }
        
        return al;
    }
}

// @saorav21994
// TC : O(n^2)  
// SC : O(n^2) 

class Solution {
    public List<List<Integer>> shiftGrid(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;
        int ele = m * n;
        k = k % ele;
        
        List<List<Integer>> res = new ArrayList<>();
        
        for (int i = 0; i < m; i++) {
            List<Integer> sublist = new ArrayList<>();
            for (int j = 0; j < n; j++) {
                int idx = findShiftedIndex(i, j, n, k, ele);
                sublist.add(grid[idx/n][idx%n]);
            }
            res.add(sublist);
        }
        
        return res;
    }
    
    public int findShiftedIndex(int i, int j, int col, int k, int ele) {
        return (i * col + j - k + ele) % ele;
    }
}

File: April2022/Kotlin/Shift2DGrid.kt
// Author: Shobhit Behl (LC: shobhitbruh)
class Solution {
    fun shiftGrid(grid: Array<IntArray>, k: Int): List<List<Int>> {
         var k = k
        val n = grid.size
        val m: Int = grid[0].size
        val li = LinkedList<Int>()
        for (i in 0 until n) {
            for (j in 0 until m) {
                li.addLast(grid[i][j])
            }
        }
        while (k-- > 0) {
            val f = li.removeLast()
            li.addFirst(f)
        }
        val al: MutableList<List<Int>> = ArrayList()
        for (i in 0 until n) {
            val g: MutableList<Int> = ArrayList()
            for (j in 0 until m) {
                g.add(li.removeFirst())
            }
            al.add(g)
        }
        return al
    }
}